### PR TITLE
apiserver: refactor graceful termination logic

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
- refactor graceful termination logic so we can write unit tests to assert on the expected behavior.
- add unit test(s) that assert on the expected behavior

It paves the way for https://github.com/kubernetes/kubernetes/pull/101257

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tkashem** [comment]
ran this test using `stress`, so far looks like there is no flake. 

```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress ./server.test -test.run=TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled
...
19m50s: 7014 runs so far, 0 failures
19m55s: 7042 runs so far, 0 failures
20m0s: 7074 runs so far, 0 failures
20m5s: 7102 runs so far, 0 failures
20m10s: 7130 runs so far, 0 failures
```

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 1]
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
+	// Start the audit backend before any request comes in. This means we must call Backend.Run
+	// before http server start serving. Otherwise the Backend.ProcessEvents call might block.
+	if s.AuditBackend != nil {
+		if err := s.AuditBackend.Run(events.WithName(InFlightRequestsDrained).Signaled()); err != nil {
+			return fmt.Errorf("failed to run the audit backend: %v", err)
+		}
+	}
+
+	serverShutdownCh, listenerStoppedCh, err := s.NonBlockingRun(events.WithName(AfterShutdownDelayDuration).Signaled())
 	if err != nil {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:381]: if `NonBlockingRun` returns an `error` we won't:
- stop the AuditBacked
- stop the `goroutine 1`
- stop the `goroutine 2`

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 2]
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
 
+	klog.V(1).Info("[graceful-termination] waiting for shutdown to be initiated")
 	<-stopCh
 
 	// run shutdown hooks directly. This includes deregistering from the kubernetes endpoint in case of kube-apiserver.
 	err = s.RunPreShutdownHooks()
 	if err != nil {
 		return err
 	}
+	klog.V(1).Info("[graceful-termination] RunPreShutdownHooks has completed")
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:397]: should we convert to structured logging?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 3]
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
 		// As soon as shutdown is initiated, /readyz should start returning failure.
-		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
-		// and stop sending traffic to this server.
-		close(s.readinessStopCh)
+		// This gives the load balancer a window defined by ShutdownDelayDuration to
+		// detect that /readyz is red and stop sending traffic to this server.
+		defer close(s.readinessStopCh)
 
-		time.Sleep(s.ShutdownDelayDuration)
+		<-events.WithName(ShutdownInitiated).Signaled()
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:369]: mixing `the new abstraction` and `regular chan` makes it a bit hard to read

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 4]
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:343]: `events` implies the usage of the regular events, rename to `signals`?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 5]
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
-		time.Sleep(s.ShutdownDelayDuration)
+		<-events.WithName(ShutdownInitiated).Signaled()
 	}()
 
-	// close socket after delayed stopCh
-	stoppedCh, err := s.NonBlockingRun(delayedStopCh)
+	// Start the audit backend before any request comes in. This means we must call Backend.Run
+	// before http server start serving. Otherwise the Backend.ProcessEvents call might block.
+	if s.AuditBackend != nil {
+		if err := s.AuditBackend.Run(events.WithName(InFlightRequestsDrained).Signaled()); err != nil {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:375]: I think that `InFlightRequestsDrained` doesn't guarantee that all in-flight requests have been processed

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 6]
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
-
-		<-stopCh
-
 		// As soon as shutdown is initiated, /readyz should start returning failure.
-		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
-		// and stop sending traffic to this server.
-		close(s.readinessStopCh)
+		// This gives the load balancer a window defined by ShutdownDelayDuration to
+		// detect that /readyz is red and stop sending traffic to this server.
+		defer close(s.readinessStopCh)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:367]: we are not interested in knowing that `close(readinessStopCh)` was closed?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 7]
```diff
@@ -330,80 +340,82 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
 // Run spawns the secure http server. It only returns if stopCh is closed
 // or the secure port cannot be listened on initially.
 func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
-	delayedStopCh := make(chan struct{})
+	events := s.TerminationEvents
+
+	signalFn := s.overrideTerminationEventSignalFunc
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:345]: so, looking at the previous code https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go it seems like knowing when `delayedStopCh` and `drainedCh` were closed would allow to write unit tests and cover the entire `Run` method.

I'm just thinking if we could restore the previous version of the code and inject a channel factory instead.
The factory would allow for creating a new chan and wrapping already existing ones. 

Then we would change 
the `Run` method to accept the factory `Run(factoryCh chanFactory)`.
`make(chan struct{})` to `factoryChan.new("name")`
`close(...)` to `chanWrapperInstance.close()`
reading from a chan `<-` to calling the `wait` method



```
type chanFactory interface {
 new(name) chanWrapper
 wrap(chan) chanWrapper
}


type chanWrapper interface {
 wait() // blocking 
 close()
 rawChan()
}
```

After that we would get:
```

func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}, factoryCh chanFactory) error {
	delayedStopCh := factoryCh.new("delayedStopCh")
         stopCh = factoryCh.wrap(stopCh)

	go func() {
		deferdelayedStopCh.close()

		stopCh.wait()

		// As soon as shutdown is initiated, /readyz should start returning failure.
		// This gives the load balancer a window defined by ShutdownDelayDuration to detect that /readyz is red
		// and stop sending traffic to this server.
		close(s.readinessStopCh)

		time.Sleep(s.ShutdownDelayDuration)
	}()

	// close socket after delayed stopCh
	stoppedCh, err := s.NonBlockingRun(delayedStopCh.rawChan())
	if err != nil {
		return err
	}

         stoppedCh = factoryCh.wrap(stoppedCh)

	drainedCh := factoryCh.new("drainedCh")
	go func() {
		defer drainedCh.close()

		// wait for the delayed stopCh before closing the handler chain (it rejects everything after Wait has been called).
		delayedStopCh.wait()

		// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
		s.HandlerChainWaitGroup.Wait()
	}()

	stopCh.wait()

	// run shutdown hooks directly. This includes deregistering from the kubernetes endpoint in case of kube-apiserver.
	err = s.RunPreShutdownHooks()
	if err != nil {
		return err
	}

	// Wait for all requests in flight to drain, bounded by the RequestTimeout variable.
	drainedCh.wait()
	// wait for stoppedCh that is closed when the graceful termination (server.Shutdown) is finished.
	stoppedCh.wait()

	return nil
}
```

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:345]: not sure how hard would it be to create these abstractions.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 1 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:379]: we are aborting, the primary goroutine that is executing `Run` will abort and all other go routines will be aborted with it. I don't think we need to be graceful when `Run` returns an error, thoughts?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 2 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:395]: `[graceful-termination]` is just a prefix so we can search all graceful termination related log entries with this keyword. Is there a way to add a prefix in structured logging?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 3 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: yes, i agree. I simplified a bit, let me know if it reads better. I guess this is a tradeoff, we need the abstraction for a flake free test, on the other hand, we lose nice idiomatic inline channel and go func.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 4 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: i renamed to `TerminationSignal` and brought back the old variable names.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 5 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: This is in keeping with the current behavior - https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L419-L420

I think this is correct - `WithWaitGroup` precedes `WithAudit` - so existing requests in flight have already been added to the audit backend. so `auditBackend.Run` will stop as soon `WithWaitGroup` has finished executing all existing requests in flight, thoughts?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 6 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: I removed `readinessStopCh` from `GenericAPIServer`. we no longer need it, we can use `ShutdownInitiated` signal instead.
```
	// as soon as shutdown is initiated, readiness should start failing
	readinessStopCh := s.TerminationSignals.ShutdownInitiated.Signaled()
	err := s.addReadyzShutdownCheck(readinessStopCh)
```

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 7 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: A couple of notes:
- i have a followup PR where i have a filter that will need the termination event channel, so we can't create these channels inside the `Run` function, see here https://github.com/kubernetes/kubernetes/pull/101257/files#diff-207875acc7ee03f987aa626f53742c02323ed8f434fd40dfd929dffb23d3f1deR789.
```
	handler = genericfilters.WithHSTS(handler, c.HSTSDirectives)
	if c.KeepListeningDuringGracefulTermination {
		handler = genericfilters.WithNotAcceptingNewRequest(handler, c.TerminationEvents.WithName(NotAcceptingNewRequest).Wait())
	}
	handler = genericfilters.WithHTTPLogging(handler)
``

-

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: Also, 
```
type chanWrapper interface {
 wait() // blocking 
 close()
 rawChan()
}
```

is similar to what we have in this PR
```
type TerminationSignal interface {
	Name() string
	Signal()
	Signaled() <-chan struct{}
}
```

- `wait` is similar to `Signaled` (except `Signaled` returns a `<- chan struct{}` as opposed to `wait` which blocks ) `Signaled` seems more idiomatic and more composable, 
- with `Signaled` we don't need `rawChan`
- `close` = `Signal`

If it makes it easier to read i can go with
```
type TerminationSignal interface {
	Name() string
	close()
	done() <-chan struct{}
}
```
Thoughts?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 1 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:372]: okay, I moved `s.AuditBackend.Run` to its original site, now it is in keeping with the current implementation

**Comment by tkashem** [comment]
the test looks good, no flake seen in about `45m` run
```
$ go test -race k8s.io/apiserver/pkg/server -c
$ stress ./server.test -test.run=TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled
47m25s: 15330 runs so far, 0 failures
47m30s: 15358 runs so far, 0 failures
47m35s: 15386 runs so far, 0 failures
47m40s: 15410 runs so far, 0 failures
...

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 8]
```diff
@@ -219,6 +219,10 @@ type Config struct {
 	// RequestWidthEstimator is used to estimate the "width" of the incoming request(s).
 	RequestWidthEstimator flowcontrolrequest.WidthEstimatorFunc
 
+	// TerminationSignals provides access to the various shutdown signals
+	// that happen during the graceful termination of the apiserver.
+	TerminationSignals *TerminationSignals
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:224]: why is this a pointer? Is it optional?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 9]
```diff
@@ -213,6 +210,10 @@ type GenericAPIServer struct {
 
 	// Version will enable the /version endpoint if non-nil
 	Version *version.Info
+
+	// TerminationSignals provides access to the various termination
+	// signals that happen during the shutdown period of the apiserver.
+	TerminationSignals *TerminationSignals
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:216]: is this ever nil?

File: staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+	Name() string
+
+	// Signal signals the event, indicating that the event has occurred.
+	// Signal is idempotent, once signaled the event stays signaled and
+	// it immediately unblocks any goroutine waiting for this event.
+	Signal()
+
+	// Signaled returns a channel that is closed when the underlying termination
+	// event has been signaled. Successive calls to Signaled return the same value.
+	Signaled() <-chan struct{}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:75]: I like this type.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 9 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:216]: does this have to be public ?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 11]
```diff
@@ -390,20 +404,22 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}) (<-chan
...
 
 	// Use an internal stop channel to allow cleanup of the listeners on error.
 	internalStopCh := make(chan struct{})
 	var stoppedCh <-chan struct{}
+	var listenerStoppedCh <-chan struct{}
 	if s.SecureServingInfo != nil && s.Handler != nil {
 		var err error
-		stoppedCh, err = s.SecureServingInfo.Serve(s.Handler, s.ShutdownTimeout, internalStopCh)
+		klog.V(1).Infof("[graceful-termination] ShutdownTimeout=%s", s.ShutdownTimeout)
+		stoppedCh, listenerStoppedCh, err = s.SecureServingInfo.Serve(s.Handler, s.ShutdownTimeout, internalStopCh)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:418]: all the changes around listenerStoppedCh might deserve its own commit. This changes the wiring.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 12]
```diff
@@ -330,58 +334,68 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
 
 		// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
 		s.HandlerChainWaitGroup.Wait()
 	}()
 
+	httpServerStoppedListeningCh := s.TerminationSignals.HTTPServerStoppedListening
+	go func() {
+		<-listenerStoppedCh
+		httpServerStoppedListeningCh.Signal()
+	}()
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:374]: move this below the block where listenerStoppedCh is returned.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go [hunk 13]
```diff
@@ -330,58 +334,68 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
+		<-delayedStopCh.Signaled()
 
 		// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
 		s.HandlerChainWaitGroup.Wait()
 	}()
 
+	httpServerStoppedListeningCh := s.TerminationSignals.HTTPServerStoppedListening
+	go func() {
+		<-listenerStoppedCh
+		httpServerStoppedListeningCh.Signal()
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:373]: do we want to log when the socket was closed ?

File: staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+func (e *namedChannelWrapper) Name() string {
+	return string(e.name)
+}
+
+func (e *namedChannelWrapper) Signal() {
+	select {
+	case <-e.ch:
+		// already closed, don't close again.
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:135]: I would remove it and let in `panic` just like normal channels

File: staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+	}
+}
+
+type namedChannelWrapper struct {
+	name string
+	ch   chan struct{}
+}
+
+func (e *namedChannelWrapper) Name() string {
+	return string(e.name)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:129]: returning `e.name` is not enough?

File: staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+		ShutdownInitiated:          newNamedChannelWrapper("ShutdownInitiated"),
+		AfterShutdownDelayDuration: newNamedChannelWrapper("AfterShutdownDelayDuration"),
+		InFlightRequestsDrained:    newNamedChannelWrapper("InFlightRequestsDrained"),
+		HTTPServerStoppedListening: newNamedChannelWrapper("HTTPServerStoppedListening"),
+	}
+}
+
+func newNamedChannelWrapper(name string) TerminationSignal {
+	return &namedChannelWrapper{
+		name: name,
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:118]: isn't this used only in the tests? if yes then I'm wondering if it could be removed.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+		doneCh: make(chan struct{}),
+	}
+}
+
+func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled(t *testing.T) {
+	s := newGenericAPIServer(t)
+
+	// record the termination events in the order they are signaled
+	var eventOrderLock sync.Mutex
+	eventOrderGot := make([]string, 0)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:113]: nit rename to `signalOrderGot` ?

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go [hunk 18]
```diff
@@ -0,0 +1,399 @@
...
+		inFlightResultCh <- resultGot
+	}()
+	select {
+	case <-inFlightStartedCh:
+	case <-time.After(5 * time.Second):
+		t.Fatalf("Waited for 5s for the in-flight request to reach the server")
+	}
+
+	// step 2: signal termination event: initiate a shutdown
+	close(stopCh)
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:181]: +1

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go [hunk 19]
```diff
@@ -0,0 +1,399 @@
...
+		// Before AfterShutdownDelayDuration event is signaled, the test
+		// will send request(s) to assert on expected behavior.
+		if e.Name() == "AfterShutdownDelayDuration" && before {
+			// it unblocks the verification step and waits for it to complete
+			<-delayedStopVerificationStep.done()
+		}
+	}
+
+	// wrap the termination signals of the GenericAPIServer so the test can inject its own callback
+	wrapTerminationSignals(t, s.TerminationSignals, func(before bool, e TerminationSignal) {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go:155]: great, so now the test simpy catches events from the `Run` method

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 8 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: it is never optional, i removed the pointer.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 9 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: it should not be public, i made it private

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 11 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:417]: i put this change in a new commit

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go (hunk 12 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go]: done

File: staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go (hunk 14 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:134]: if `Signal` is called more than once then the test would fail, the test actually verifies that the events are signaled exactly once and in the right order:
```
	terminationSignalOrderExpected := []string{
		string("ShutdownInitiated"),
		string("AfterShutdownDelayDuration"),
		string("HTTPServerStoppedListening"),
		string("InFlightRequestsDrained"),
	}
```

File: staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go (hunk 15 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go]: fixed

File: staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go (hunk 16 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:117]: yes, it can be removed, it's just easier for the the test to get the name instead of using a type assertion.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apiserver/pkg/server/graceful_termination.go:117]: I would remove it and add it to the embedded type in the test.

File: staging/src/k8s.io/apiserver/pkg/server/secure_serving.go [hunk 20]
```diff
@@ -192,6 +192,67 @@ func (s *SecureServingInfo) Serve(handler http.Handler, shutdownTimeout time.Dur
...
+	return stoppedCh, err
+}
+
+// ServeWithListenerStopped runs the secure http server. It fails only if certificates cannot be loaded or the initial listen call fails.
+// The actual server loop (stoppable by closing stopCh) runs in a go routine, i.e. ServeWithListenerStopped does not block.
+// It returns a stoppedCh that is closed when all non-hijacked active requests have been processed.
+// It returns a listenerStoppedCh that is closed when the underlying http Server has stopped listening.
+// TODO: do a follow up PR to remove this function, change 'Serve' to return listenerStoppedCh
+//  and update all components that call 'Serve'
+func (s *SecureServingInfo) ServeWithListenerStopped(handler http.Handler, shutdownTimeout time.Duration, stopCh <-chan struct{}) (<-chan struct{}, <-chan struct{}, error) {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/apiserver/pkg/server/secure_serving.go:206]: we are going to remove the `Serve` method in a follow-up PR.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/secure_serving.go:206]: yes, i have added a TODO for that as well
